### PR TITLE
Issue #888 - add support to validate extensible and preferred binding

### PR DESCRIPTION
**Pull request by JohnTimm**

### CODE REVIEW

File: fhir-ig-carin-bb/src/test/resources/JSON/EOB1.json [hunk 1]
```diff
@@ -0,0 +1,1101 @@
+{
```
Comment (lmsurpre) [inline fhir-ig-carin-bb/src/test/resources/JSON/EOB1.json:1]: do we make the examples accessible from the registry, or we need it in the test dir too?

File: fhir-path/src/main/java/com/ibm/fhir/path/function/MemberOfFunction.java [hunk 2]
```diff
@@ -63,7 +67,7 @@ public int getMinArity() {
 
     @Override
     public int getMaxArity() {
-        return 1;
+        return 2;
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/MemberOfFunction.java:70]: Please update the javadoc to indicate that we've added our own flavor of 'memberOf' which allows the caller to pass a binding strength

File: fhir-ig-mcode/src/test/java/com/ibm/fhir/ig/mcode/test/ConstraintGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import com.ibm.fhir.model.annotation.Constraint;
+import com.ibm.fhir.model.resource.StructureDefinition;
+import com.ibm.fhir.model.type.Extension;
+import com.ibm.fhir.model.util.ModelSupport;
+import com.ibm.fhir.profile.ProfileSupport;
+import com.ibm.fhir.registry.resource.FHIRRegistryResource;
+import com.ibm.fhir.registry.spi.FHIRRegistryResourceProvider;
+
+public class ConstraintGeneratorTest {
+    public static void main(String[] args) {
```
Comment (lmsurpre) [inline fhir-ig-mcode/src/test/java/com/ibm/fhir/ig/mcode/test/ConstraintGeneratorTest.java:19]: should it be a proper test?

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java [hunk 4]
```diff
@@ -1457,5 +1462,59 @@ public void unsetExternalConstant(String name) {
...
+
+        /**
+         * Get the list of issues that occurred during evaluation
+         *
+         * @return
+         *     the list of issues that occurred during evaluation
+         */
+        public List<Issue> getIssues() {
+            return issues;
+        }
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1511]: is it only warnings, or might we expect errors here too?

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java [hunk 5]
```diff
@@ -161,15 +171,17 @@ private String generate(Node node) {
         }
 
         if (hasVocabularyConstraint(elementDefinition)) {
-            return generateVocabularyConstraint(elementDefinition);
+            String expr = generateVocabularyConstraint(elementDefinition);
+            if (node.children.stream().noneMatch(child -> hasConstraint(child))) {
```
Comment (lmsurpre) [inline fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:175]: maybe a comment to explain this line?  like why do we only add the vocabulary constraint if the child has no constraints.

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java [hunk 6]
```diff
@@ -161,15 +171,17 @@ private String generate(Node node) {
...
+                return expr;
+            }
+            sb.append(expr).append(" and ");
```
Comment (lmsurpre) [inline fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:178]: will there ALWAYS be another constraint following this and?

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java [hunk 7]
```diff
@@ -514,8 +534,12 @@ private boolean hasVocabularyConstraint(ElementDefinition elementDefinition) {
...
             String strength = binding.getStrength().getValue();
             String valueSet = binding.getValueSet().getValue();
-            return (!"required".equals(baseStrength) && "required".equals(strength))
-                    || ("required".equals(baseStrength) && "required".equals(strength) && !valueSetEqualsIgnoreVersion(valueSet, baseValueSet));
+            return (!"preferred".equals(baseStrength) && "preferred".equals(strength)) ||
+                    ("preferred".equals(baseStrength) && "preferred".equals(strength) && !valueSetEqualsIgnoreVersion(valueSet, baseValueSet)) ||
+                    (!"extensible".equals(baseStrength) && "extensible".equals(strength)) ||
+                    ("extensible".equals(baseStrength) && "extensible".equals(strength) && !valueSetEqualsIgnoreVersion(valueSet, baseValueSet)) ||
+                    (!"required".equals(baseStrength) && "required".equals(strength)) ||
+                    ("required".equals(baseStrength) && "required".equals(strength) && !valueSetEqualsIgnoreVersion(valueSet, baseValueSet));
```
Comment (lmsurpre) [inline fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:542]: quite a piece of logic here, maybe a comment to explain?

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java [hunk 8]
```diff
@@ -1457,5 +1462,59 @@ public void unsetExternalConstant(String name) {
...
             return externalConstantMap.containsKey(name);
         }
+
+        /**
+         * Set the constraint currently under evaluation
+         *
+         * @param constraint
+         *     the constraint currently under evaluation
+         */
+        public void setConstraint(Constraint constraint) {
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1472]: when is setting the constraint required?  when is it not required?

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java [hunk 9]
```diff
@@ -1457,5 +1462,59 @@ public void unsetExternalConstant(String name) {
...
+            this.constraint = constraint;
+        }
+
+        /**
+         * Unset the constraint currently under evaluation
+         */
+        public void unsetConstraint() {
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1479]: when is unsetting the constraint required? when is it not required?

File: fhir-ig-carin-bb/src/test/resources/JSON/EOB1.json (hunk 1 above)
Comment (JohnTimm) [reply fhir-ig-carin-bb/src/test/resources/JSON/EOB1.json:1]: We do not make examples accessible from the registry. This example did not come from the IG package. It was in the IG github repo.

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java (hunk 4 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1513]: I made it flexible enough to support both.

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java (hunk 6 above)
Comment (JohnTimm) [reply fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:179]: if hasConstraint returns true on the check above it, then yes.

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java (hunk 7 above)
Comment (JohnTimm) [reply fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java]: I have simplified this logic in the latest commit. It should be much easier to understand now.

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java (hunk 9 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1481]: The unset is driven by the Validator (after it has evaluated the current constraint and before it changes to the context constraint).

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java (hunk 8 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1474]: It is driven by the Validator to give the FHIRPathEvaluator and any FHIRPathFunction(s) access to it during evaluation. Thus far, the only function that cares about it is MemberOfFunction.

File: fhir-ig-mcode/src/test/java/com/ibm/fhir/ig/mcode/test/ConstraintGeneratorTest.java (hunk 3 above)
Comment (JohnTimm) [reply fhir-ig-mcode/src/test/java/com/ibm/fhir/ig/mcode/test/ConstraintGeneratorTest.java:19]: This is more of a diagnostic driver to print out the generated constraints for manual inspection. I don't anticipate that we will make this part of the automated unit tests for this module.

File: fhir-path/src/main/java/com/ibm/fhir/path/function/MemberOfFunction.java (hunk 2 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/MemberOfFunction.java:76]: Done

File: fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java (hunk 5 above)
Comment (JohnTimm) [reply fhir-profile/src/main/java/com/ibm/fhir/profile/ConstraintGenerator.java:176]: Added comments to clarify the behavior.

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1517]: but currently, an error will result in a thrown exception, right?  and presumably an empty list of issues?

Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java]: Added javadoc to clarify that these are "supplemental issues" used to convey additional information to the client.

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java (hunk 8 above)
Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1478]: new method doc looks good

File: fhir-path/src/main/java/com/ibm/fhir/path/function/MemberOfFunction.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/MemberOfFunction.java:76]: new class doc looks good

**Pull request merged**